src/test/java/com/alphawallet/attestation/demo/VerifyAttestation.sol

```diff
@@ -0,0 +1,793 @@
+        {
+            revert();
+        }
+        
+        //Ensure base is correct
+        if (!ecEquals(base, H))
+        {
+            revert();
+        }
+        
+        //Note: if needed we can hand optimise this, but the gain is small
+        //bytes memory cArray = abi.encodePacked(GPoint, pok.base, comPoint1, comPoint2, pok.tPoint); //
+        bytes memory cArray = concat5(GPoint, pok.base, com1, com2, pok.tPoint); 
+        uint256 c = mapToCurveMultiplier(cArray);
+        
+        lhs = ecMul(pok.challenge, base[0], base[1]);
+        if (lhs[0] == 0 && lhs[1] == 0) { revert(); } //early revert to avoid spending more gas
+        
+        //ECPoint riddle muliply by proof (component hash)
+        rhs = ecMul(c, riddle[0], riddle[1]);
+        if (rhs[0] == 0 && rhs[1] == 0) { revert(); } //early revert to avoid spending more gas
+        
+        //ecVals[2] = rhs[0];
+        //ecVals[3] = rhs[1];
+        //ecVals[4] = c;
+        
+        //Add result of riddle.multiply(c) to point
+        uint256[2] memory tPointCoords;
+        (tPointCoords[0], tPointCoords[1]) = extractXYFromPoint(pok.tPoint);
+        rhs = ecAdd(rhs, tPointCoords);
+        
+        //ecVals[0] = lhs[0];
+        //ecVals[1] = lhs[1];
+        
+        //ecVals[2] = rhs[0];
+        //ecVals[3] = rhs[1];
+        
+        return ecEquals(lhs, rhs);
+    }
+    
+    function ecEquals(uint256[2] memory ecPoint1, uint256[2] memory ecPoint2) private view returns(bool)
+    {
+        return (ecPoint1[0] == ecPoint2[0] && ecPoint1[1] == ecPoint2[1]);
+    }
+
+    function decodeAttestation(bytes memory attestation) public view returns(bool)
+    {
+        Length memory len;
+        Exponent memory pok;
+        OctetString memory octet;
+
+        require(attestation[0] == (CONSTRUCTED_TAG | SEQUENCE_TAG));
+
+        len = decodeLength(attestation, 1);
+        octet.decodeIndex = len.decodeIndex;
+
+        //decode parts
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.base = octet.byteCode;
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.riddle = octet.byteCode;
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.challenge = bytesToUint(octet.byteCode);
+        octet = decodeOctetString(attestation, octet.decodeIndex);
+        pok.tPoint = octet.byteCode;
+
+        //Calculate LHS ECPoint
+        (uint256 x, uint256 y) = extractXYFromPoint(pok.base);
+        uint256[2] memory lhs = ecMul(pok.challenge, x, y);
+        
+        if (lhs[0] == 0 && lhs[1] == 0) { return false; } //early revert to avoid spending more gas
+
+        //Calculate RHS ECPoint
+        (x, y) = extractXYFromPoint(pok.riddle);
+        
+        //The hand optimised concat4 is more optimal than using abi.encodePacked (checked the gas costs of both)
+        bytes memory cArray = concat4(GPoint, pok.base, pok.riddle, pok.tPoint);
```

Comment:
We strictly don't need the G point. It can be removed and will be as part of issue #72